plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -221,6 +224,45 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
         }
     }
 
+    @Override
+    public JdbcTableHandle getTableHandle(ConnectorSession session, PreparedQuery preparedQuery)
+    {
+        ImmutableList.Builder<JdbcColumnHandle> columns = ImmutableList.builder();
+        try (Connection connection = connectionFactory.openConnection(session);
+                PreparedStatement preparedStatement = queryBuilder.prepareStatement(this, session, connection, preparedQuery)) {
+            ResultSetMetaData metaData = preparedStatement.getMetaData();
+            if (metaData == null) {
+                throw new UnsupportedOperationException("ResultSetMetaData not provided for query");
+            }
+            for (int column = 1; column <= metaData.getColumnCount(); column++) {
+                JdbcTypeHandle jdbcTypeHandle = new JdbcTypeHandle(
+                        metaData.getColumnType(column),
+                        Optional.ofNullable(metaData.getColumnTypeName(column)),
+                        Optional.of(metaData.getPrecision(column)),
+                        Optional.of(metaData.getScale(column)),
+                        Optional.empty(), // TODO support arrays
+                        Optional.of(metaData.isCaseSensitive(column) ? CASE_SENSITIVE : CASE_INSENSITIVE));
+                Type type = toColumnMapping(session, connection, jdbcTypeHandle)
+                        .orElseThrow(() -> new IllegalArgumentException("Unsupported type: " + jdbcTypeHandle))
+                        .getType();
+                columns.add(new JdbcColumnHandle(metaData.getColumnName(column), jdbcTypeHandle, type));
+            }
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+
+        return new JdbcTableHandle(
+                new JdbcQueryRelationHandle(preparedQuery),
+                TupleDomain.all(),
+                ImmutableList.of(),
+                Optional.empty(),
+                OptionalLong.empty(),
+                Optional.of(columns.build()),
+                ImmutableSet.of(), // TODO return other tables referenced by the query
```

Comment:
We won't be able to do this, since query is opaque.

Instead of TODO add a `// Note` that we're not doing that.

AFAIR, this field is used for cache eviction. I think we should make `io.trino.plugin.jdbc.JdbcTableHandle#otherReferencedTables` and `Optional` field (so that we can mark "unknown" state) so cache eviction remains correct

Reply:
I filed an issue https://github.com/trinodb/trino/issues/12526, and added a link in the code along with the not.

Reply:
Thanks, fine to do it as a follow-up.